### ISSUE TITLE
Benchmarking takes ages to generate files

### ISSUE THREAD

**Issue by tillea** [MikkelSchubert/adapterremoval#9]
Hi,

I'm a member of the Debian Med team which is a group inside Debian to package free software in life sciences and medicine for official Debian. I'd like to inform you that I have packaged adpaterremoval and it is now available in Debian unstable.  I also tried to experiment with the benchmarking which is really interesting since we could on one hand test and on the other hand compare several of the tools we have packaged (not all tools in the suite are packaged yet, but a fair amount which makes a comparison interesting).

I noticed that you are using pIRS for creating artificial sequences.  When having a packaging centric look into pIRS I noticed that some files are copyrighted by Illumina under a non-free / non-distributable license (see [debian/copyright](https://anonscm.debian.org/cgit/debian-med/pirs.git/tree/debian/copyright))  I might try to contact the pIRS authors or even Illumina about a clarification but I doubt that chances are very good.  May be its easier to replace the sequence creation program at all.  For instance in Debian we have packaged [art-nextgen-simulation-tools](https://blends.debian.org/med/tasks/bio#art-nextgen-simulation-tools) and [artfastgenerator](https://blends.debian.org/med/tasks/bio#artfastqgenerator).  I personally can not tell about their quality and how these compare to pIRS - I just know that there might be potential replacements for pIRS (may be even more?).

Another issue I have with the benchmarking suite is that I takes ages to finalise the generation process filling up the hard disk with GBytes of files.  I admit I have interrupted the process several times - even after trying

`--- a/benchmark/benchmark.sh
+++ b/benchmark/benchmark.sh
@@ -44,7 +44,7 @@ ADAPTER_ID_INSERT_SIZES=($(seq 250 5 350
 MAX_THREADS=4

 # Number of read (pairs) to simulate using pIRS for each replicate
-SIMULATED_NREADS=1000000
+SIMULATED_NREADS=100

 results=$(mktemp -d /tmp/adapter-benchmark.XXXXXX)
 echo "**\* Results will be found in $results ***"
`
to reduce the size.

May be that's due to the fact that I used the _unpatched_ pIRS.  Could you please clarify the reason for the pIRS patch.  I could add this to my (inofficial / non-distributable) packaging for pIRS to enable a sensible benchmarking.

Kind regards

```
    Andreas.
```

**Comment by MikkelSchubert** [comment]
Hey Andreas,
First off, thank you very much for adding AdapterRemoval to Debian unstable!

With regards to the benchmarking script, let me warn you that I included this for the sake of reproducability of the published results. While I hope that you find it useful, it had already reached the point where I was considering rewriting the entire thing in a a more suitable language than Bash.

Also, you need to be be aware that I relied on making modifications to several of the programs used, including pIRS, as you've noticed. I did this mainly for the purpose of allowing a meaningful run-time comparisons, by disabling mandatory compression, but also to tag the output of some programs (for merged reads), in order ensure that I could easily count those correctly. The programs patched are listed in the README in the benchmark folder, and the patches are in the patches folder:

  - leeHom: Disable gzip-compression and mark merged reads with "M_",
    to simplify analysis.
  - PEAR: Marks merged reads with "M_", to simplify processing.
  - pIRS: Adds support for user supplied adapter sequences and barcodes, allow
    zero length inserts.

I selected pIRS because it was capable of using empirical error rates for simulated Illumina data. This may not be necessary for your purposes, but it is worthwhile to keep in mind. However, I had to patch pIRS to allow it to simulate DNA fragments shorter than the read-length (by default, pIRS simply drops such inserts), which are then padded to the correct length using the user-defined adapter sequences. Without that change, there are no adapter sequences to trim! I also allowed for multiple adapter pairs to be specified, in which case pIRS would select between these per simulated read pair. I further patched it to let me simulate a fixed number of reads (rather than by coverage), as adding barcodes to reads (not used in the benchmark).

There is also a newer version of pIRS available, with a much improved code-base, but I also discovered that after everything was already said and done. If that version lacks the problematic files, then you could perhaps adapt the patches to it, so it can be used instead: https://github.com/galaxy001/pirs

Finally, with regards to disk usage, it is indeed the case that the script will use a LOT of disk-space. All simulated reads are generated up-front. The script counting the trimmed reads will delete the results files from each program straight after they've been produced, but other than that I did not attempt to optimize disk usage. I intentionally did not generate compressed reads, since support for reading gzipped reads was uneven among the programs I tested. For the publication, I simply used an 256 Gb SDD drive, that I had formatted for this purpose. You should, however, be able to reduce the amount of space used by decreasing NUM_REPLICATES, by not simulating both 100 and 200 bp long reads (modify READ_LENGTHS), and by decreasing SIMULATED_NREADS, as you've already done.

And you'll probably want to make sure that you are not generating data for the adapter_id step (comment out the calls to benchmark_adapter_id). You might also want to skip the 'mixed' simulations, in which a single data-set contains reads with different adapter pairs.

Cheers,
 Mikkel

**Comment by tillea** [comment]
Hi Mikkel,

On Mon, Aug 08, 2016 at 05:09:14AM -0700, Mikkel Schubert wrote:
> First off, thank you very much for adding AdapterRemoval to Debian unstable!

You are welcome. :-)

> With regards to the benchmarking script, let me warn you that I included this for the sake of reproducability of the published results. While I hope that you find it useful, it had already reached the point where I was considering rewriting the entire thing in a a more suitable language than Bash.

I could imagine that Bash at some points might reach its limits.
 
> Also, you need to be be aware that I relied on making modifications to several of the programs used, including pIRS, as you've noticed. I did this mainly for the purpose of allowing a meaningful run-time comparisons, by disabling mandatory compression, but also to tag the output of some programs (for merged reads), in order ensure that I could easily count those correctly. The programs patched are listed in the README in the benchmark folder, and the patches are in the patches folder:
> 
>   - leeHom: Disable gzip-compression and mark merged reads with "M_",
>     to simplify analysis.
>   - PEAR: Marks merged reads with "M_", to simplify processing

Thanks for these hints bit these two programs are not yet packaged (even
if it might be interesting packaging targets as well).

>   - pIRS: Adds support for user supplied adapter sequences and barcodes, allow
>     zero length inserts.
> 
> I selected pIRS because it was capable of using empirical error rates for simulated Illumina data. This may not be necessary for your purposes, but it is worthwhile to keep in mind. However, I had to patch pIRS to allow it to simulate DNA fragments shorter than the read-length (by default, pIRS simply drops such inserts), which are then padded to the correct length using the user-defined adapter sequences. Without that change, there are no adapter sequences to trim! I also allowed for multiple adapter pairs to be specified, in which case pIRS would select between these per simulated read pair. I further patched it to let me simulate a fixed number of reads (rather than by coverage), as adding barcodes to reads (not used in the benchmark).

Did you ever considered to forward this change to pIRS upstream since it
might be of general interest.  Since I now learned that pIRS maintenance
is continued this might be worth the effort.

> There is also a newer version of pIRS available, with a much improved code-base, but I also discovered that after everything was already said and done. If that version lacks the problematic files, then you could perhaps adapt the patches to it, so it can be used instead: https://github.com/galaxy001/pirs

Good hint.  Unfortunately the files remain inside the source but the
information that its on Github and not on some possibly unmaintained ftp
archive with no contact information is worth trying to opening an issue.
Very helpful. - I'll keep you updated whether pIRS can be released
inside Debian and if possible I'll add your patch.
 
> Finally, with regards to disk usage, it is indeed the case that the script will use a LOT of disk-space. All simulated reads are generated up-front. The script counting the trimmed reads will delete the results files from each program straight after they've been produced, but other than that I did not attempt to optimize disk usage. I intentionally did not generate compressed reads, since support for reading gzipped reads was uneven among the programs I tested. For the publication, I simply used an 256 Gb SDD drive, that I had formatted for this purpose. You should, however, be able to reduce the amount of space used by decreasing NUM_REPLICATES, by not simulating both 100 and 200 bp long reads (modify READ_LENGTHS), and by decreasing SIMULATED_NREADS, as you've already done.
> 
> And you'll probably want to make sure that you are not generating data for the adapter_id step (comment out the calls to benchmark_adapter_id). You might also want to skip the 'mixed' simulations, in which a single data-set contains reads with different adapter pairs.

I'll check these helpful hints once I've got pIRS 2.0.1 packaged.

Thanks for your helpful hints

       Andreas.

-- 
http://fam-tille.de

**Comment by MikkelSchubert** [comment]
Hey Andreas,

> Did you ever considered to forward this change to pIRS upstream since it
might be of general interest.  Since I now learned that pIRS maintenance
is continued this might be worth the effort.

I went back and checked, and unfortunately the large scale of changes to the code-base between pIRS v1 and v2 would mean that the patch would have to be redone from scratch.  Furthermore, I implemented the generation of exactly N reads by simply ripping out the existing code, which calculates the number of reads to generate based on a desired depth of coverage, so this would have to be re-added in a way that does not break existing functionality. I have to admit that (re)doing this is very low on my list of priorities. 

Cheers,
 Mikkel
